wrappers/binaries.go

```diff
@@ -33,100 +33,82 @@ import (
 }
 
-func generateSnapBinaryWrapper(app *snap.AppInfo, pkgPath string) (string, error) {
+func generateSnapBinaryWrapper(app *snap.AppInfo) (string, error) {
 	wrapperTemplate := `#!/bin/sh
 set -e
 
 # snap info
-{{.NewAppVars}}
+{{.EnvVars}}
 
 if [ ! -d "$SNAP_USER_DATA" ]; then
    mkdir -p "$SNAP_USER_DATA"
 fi
 export HOME="$SNAP_USER_DATA"
 
-# Snap name is: {{.SnapName}}
-# App name is: {{.AppName}}
+# Snap name is: {{.App.Snap.Name}}
+# App name is: {{.App.Name}}
 
-ubuntu-core-launcher {{.UdevAppName}} {{.AaProfile}} {{.Target}} "$@"
+{{.App.Invocation}} "$@"
 `
 
 	if err := snap.ValidateApp(app); err != nil {
 		return "", err
 	}
 
-	actualBinPath := binPathForBinary(pkgPath, app)
-
 	var templateOut bytes.Buffer
 	t := template.Must(template.New("wrapper").Parse(wrapperTemplate))
 	wrapperData := struct {
-		SnapName    string
-		AppName     string
-		SnapArch    string
-		SnapPath    string
-		Version     string
-		Revision    int
-		UdevAppName string
-		Home        string
-		Target      string
-		AaProfile   string
-		OldAppVars  string
-		NewAppVars  string
+		App     *snap.AppInfo
+		EnvVars string
+		// XXX: needed by snapenv
+		SnapName string
+		SnapArch string
+		SnapPath string
+		Version  string
+		Revision int
+		Home     string
 	}{
-		SnapName:    app.Snap.Name(),
-		AppName:     app.Name,
-		SnapArch:    arch.UbuntuArchitecture(),
-		SnapPath:    pkgPath,
-		Version:     app.Snap.Version,
-		Revision:    app.Snap.Revision,
-		UdevAppName: app.SecurityTag(),
-		Home:        "$HOME",
-		Target:      actualBinPath,
-		AaProfile:   app.SecurityTag(),
+		App: app,
+		// XXX: needed by snapenv
+		SnapName: app.Snap.Name(),
+		SnapArch: arch.UbuntuArchitecture(),
+		SnapPath: app.Snap.MountDir(),
+		Version:  app.Snap.Version,
+		Revision: app.Snap.Revision,
+		Home:     "$HOME",
 	}
 
-	newVars := []string{}
+	envVars := []string{}
 	for _, envVar := range append(
 		snapenv.GetBasicSnapEnvVars(wrapperData),
 		snapenv.GetUserSnapEnvVars(wrapperData)...) {
-		newVars = append(newVars, quoteEnvVar(envVar))
+		envVars = append(envVars, quoteEnvVar(envVar))
 	}
-	wrapperData.NewAppVars = strings.Join(newVars, "\n")
+	wrapperData.EnvVars = strings.Join(envVars, "\n")
 
 	t.Execute(&templateOut, wrapperData)
 
 	return templateOut.String(), nil
 }
 
-func addPackageBinaries(s *snap.Info) error {
+// AddSnapBinaries writes the wrapper binaries for the applications from the snap which aren't services.
+func AddSnapBinaries(s *snap.Info) error {
 	if err := os.MkdirAll(dirs.SnapBinariesDir, 0755); err != nil {
 		return err
 	}
 
-	baseDir := s.MountDir()
-
 	for _, app := range s.Apps {
 		if app.Daemon != "" {
 			continue
 		}
 
-		// this will remove the global base dir when generating the
-		// service file, this ensures that /snap/foo/1.0/bin/start
-		// is in the service file when the SetRoot() option
-		// is used
-		realBaseDir := stripGlobalRootDir(baseDir)
```

Comment:
It is important that this stripGlobalRootDir() is performed somewhere. Without that u-d-f will no longer be able to install snaps that start because the globalRootDir is prefixed in e.g. the service files Exec= section. AFAICS the replacement is Placement() which uses app.MountDir() which uses dirs.SnapSnapsDir which will have the prefix.